backend/session/libseat.c

```diff
@@ -0,0 +1,211 @@
+#include <wlr/config.h>
+#include <wlr/util/log.h>
+#include "util/signal.h"
+
+#include <libseat.h>
+
+const struct session_impl session_libseat;
+
+struct libseat_device {
+	struct wl_list link;
+	int fd;
+	int device_id;
+};
+
+struct libseat_session {
+	struct wlr_session base;
+
+	struct libseat *seat;
+	struct wl_event_source *event;
+	struct wl_list devices;
+
+	bool signal;
+	bool activated;
+};
+
+static void handle_enable_seat(struct libseat *seat, void *data) {
+	struct libseat_session *session = data;
+	session->activated = true;
+	session->base.active = true;
+
+	// The session signal is muted until setup is complete, as it might not be
+	// prepared by the caller yet.
+	if (session->signal) {
+		wlr_signal_emit_safe(&session->base.session_signal, session);
+	}
+}
+
+static void handle_disable_seat(struct libseat *seat, void *data) {
+	struct libseat_session *session = data;
+	session->activated = false;
+	session->base.active = false;
+
+	// The session signal is muted until setup is complete, as it might not be
+	// prepared by the caller yet.
+	if (session->signal) {
+		wlr_signal_emit_safe(&session->base.session_signal, session);
+	}
+	libseat_disable_seat(session->seat);
+}
+
+static int libseat_event(int fd, uint32_t mask, void *data) {
+	struct libseat *seat = data;
+	libseat_dispatch(seat, 0);
+	return 1;
+}
+
+static struct libseat_seat_listener seat_listener = {
+	.enable_seat = handle_enable_seat,
+	.disable_seat = handle_disable_seat,
+};
+
+static struct libseat_session *libseat_session_from_session(
+		struct wlr_session *base) {
+	assert(base->impl == &session_libseat);
+	return (struct libseat_session *)base;
+}
+
+static struct wlr_session *libseat_session_create(struct wl_display *disp) {
+	struct libseat_session *session = calloc(1, sizeof(*session));
+	if (!session) {
+		wlr_log(WLR_ERROR, "Allocation failed: %s", strerror(errno));
+		return NULL;
+	}
+
+	wl_list_init(&session->devices);
+
+	session->seat = libseat_open_seat(&seat_listener, session);
+	if (session->seat == NULL) {
+		wlr_log(WLR_ERROR, "Unable to create seat: %s\n", strerror(errno));
+		goto error;
+	}
+
+	const char *seat_name = libseat_seat_name(session->seat);
+	if (seat_name == NULL) {
+		wlr_log(WLR_ERROR, "Unable to get seat info: %s\n", strerror(errno));
+		goto error;
+	}
+	snprintf(session->base.seat, sizeof(session->base.seat), "%s", seat_name);
+
+	struct wl_event_loop *event_loop = wl_display_get_event_loop(disp);
+	session->event = wl_event_loop_add_fd(event_loop, libseat_get_fd(session->seat),
+		WL_EVENT_READABLE, libseat_event, session->seat);
+
+	if (libseat_dispatch(session->seat, 0) == -1 && errno != EAGAIN) {
+		wlr_log(WLR_ERROR, "libseat dispatch failed: %s\n", strerror(errno));
+		goto error;
+	}
+
+	if (!session->activated) {
+		wlr_log(WLR_INFO, "Waiting for seat activation on %s", session->base.seat);
+	}
+
+	while (!session->activated) {
+		if (wl_event_loop_dispatch(event_loop, -1) == -1) {
```

Comment:
This blocks while the session is inactive on startup. Is this what other session backends do too?

Reply:
This was initially to mimick the wait from CanGraphical in logind in initial revisions, before which a few problems were present.

I'd much prefer removing it and leave activation to the signal, so I'll give that a try again.

Reply:
Seems to work - it requires moving removing the blind `session->active = true` from the caller, as a session is not guaranteed to be enabled.

Note: I moved session_signal init together with active into the backends, as libseat can raise that signal early and had a hack to mute signals until returning from create.